compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp

```diff
@@ -0,0 +1,87 @@
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-propagate-dispatch-size-bounds"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUPROPAGATEDISPATCHSIZEBOUNDSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+static void applyBounds(FunctionOpInterface funcOp,
+                        ArrayRef<int64_t> workitemCounts,
+                        ArrayRef<int64_t> workgroupCounts) {
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    if (auto tidOp = dyn_cast<gpu::ThreadIdOp>(op)) {
+      // Not the actual code, pending an upstream merge.
+      tidOp.setUpperBoundAttr(b.getIndexAttr(
+          workitemCounts[static_cast<uint32_t>(tidOp.getDimension())]));
+    } else if (auto wgSizeOp =
+                   dyn_cast<IREE::HAL::InterfaceWorkgroupSizeOp>(op)) {
+      wgSizeOp.setUpperBoundAttr(b.getIndexAttr(
+          workitemCounts[wgSizeOp.getDimension().getZExtValue()]));
+    } else if (auto wgIdOp = dyn_cast<IREE::HAL::InterfaceWorkgroupIDOp>(op)) {
+      wgIdOp.setUpperBoundAttr(b.getIndexAttr(
+          workgroupCounts[wgIdOp.getDimension().getZExtValue()]));
+    } else if (auto wgCountOp =
+                   dyn_cast<IREE::HAL::InterfaceWorkgroupCountOp>(op)) {
+      wgCountOp.setUpperBoundAttr(b.getIndexAttr(
+          workgroupCounts[wgCountOp.getDimension().getZExtValue()]));
+    }
+  });
+}
+
+struct GPUPropagateDispatchSizeBoundsPass final
+    : impl::GPUPropagateDispatchSizeBoundsPassBase<
+          GPUPropagateDispatchSizeBoundsPass> {
+  using GPUPropagateDispatchSizeBoundsPassBase::
+      GPUPropagateDispatchSizeBoundsPassBase;
+  void runOnOperation() override {
+    FunctionOpInterface funcOp = getOperation();
+    SmallVector<int64_t> workitemCounts(3, maxWorkgroupSize);
+    SmallVector<int64_t> workgroupCounts = {maxWorkgroupsX, maxWorkgroupsYZ,
+                                            maxWorkgroupsYZ};
+
+    std::optional<SmallVector<int64_t>> staticWorkitemCounts =
+        getWorkgroupSize(funcOp);
+    if (staticWorkitemCounts) {
+      if (staticWorkitemCounts->size() > 3) {
+        // How to flag this but not fail the pipeline?
+        LLVM_DEBUG(llvm::dbgs()
+                   << "more than 3 workgroup size dimensions in GPU code\n");
+        return;
+      }
+      llvm::copy(*staticWorkitemCounts, workitemCounts.begin());
+    }
+    SmallVector<int64_t> staticWorkgroupCounts = getStaticNumWorkgroups(funcOp);
+    if (staticWorkgroupCounts.size() > 3) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "more than 3 workgroup count dimensions in GPU code\n");
+      return;
+    }
+    for (auto [known, upperBound] :
+         llvm::zip(staticWorkgroupCounts, workgroupCounts))
+      if (known != ShapedType::kDynamic)
+        upperBound = known;
```

Comment:
Style nit: IREE style is to always add braces for nested statements, unlike LLVM.